### PR TITLE
Fix CosineAnnealingWarmRestarts resumption from high step counts

### LINKED ISSUE #88791: Potential bug in torch.optim.lr_scheduler.CosineAnnealingWarmRestarts
https://github.com/pytorch/pytorch/blob/fb5c6ae61f1f622ec388ae9fa00e7683ce1729ce/torch/optim/lr_scheduler.py#L1407

Should be self.T_cur = self.T_cur % self.T_i, not self.T_cur = self.T_cur - self.T_i.

### PR DESCRIPTION
**Pull request by Birch-san**
Fixes #88791.

If you try to resume a CosineAnnealingWarmRestarts schedule from anything except a low step count, you get fun learning rates that make your day better:  
image

See the [notebook](https://github.com/Birch-san/scheduler-play/blob/65eafcd7264042d16a8a295c84c219736c223131/script/repro.ipynb) with which I classified the behaviour and manually validated the fix.

image

It's worth noting though that #88791 proposes a loopless solution based on modular arithmetic. I haven't checked whether the solutions are equivalent, but that approach sounds better if it can be confirmed as working.

### CODE REVIEW

**Comment by neverix** [comment]
[I can confirm that they are the same](https://colab.research.google.com/drive/1Tv3EOkZ7uPAjazAxNvbjjB9O2BnanANI?usp=sharing)

**Comment by janeyx99** [comment]
Thank you for the fix. I am currently considering accepting https://github.com/pytorch/pytorch/pull/110497 over this one as it removes a branch of code that one would need to consider and seems equivalent. I will close this PR for now, but if I am wrong about equivalency, I will reopen this PR.
